azure-pipelines.yml

```diff
@@ -29,34 +30,35 @@ stages:
             clean: true
             submodules: false
           - script: |
-              if exist "$(buildOutputLocation)" (
-                rmdir /S /Q "$(buildOutputLocation)"
-              )
               call "%ProgramFiles%\Microsoft Visual Studio\2022\Preview\Common7\Tools\VsDevCmd.bat" ^
-              -host_arch=amd64 -arch=amd64 -no_logo
-              cmake -G Ninja -DCMAKE_CXX_COMPILER=cl -DCMAKE_BUILD_TYPE=Release ^
-              -S $(Build.SourcesDirectory)\tools -B $(buildOutputLocation)
-              cmake --build $(buildOutputLocation)
-            displayName: 'Build Support Tools'
-            env: { TMP: $(tmpDir), TEMP: $(tmpDir) }
-          - task: BatchScript@1
-            displayName: 'Enforce clang-format'
-            timeoutInMinutes: 60
+                -host_arch=amd64 -arch=amd64 -no_logo
+              cmake -G Ninja -S $(Build.SourcesDirectory)/tools/format -B $(tmpDir)/format-build
+              cmake --build $(tmpDir)/format-build
+            displayName: 'clang-format'
+            timeoutInMinutes: 5
             condition: succeededOrFailed()
-            inputs:
-              filename: 'azure-devops/enforce-clang-format.cmd'
-              failOnStandardError: true
-              arguments: '$(buildOutputLocation)/parallelize/parallelize.exe'
             env: { TMP: $(tmpDir), TEMP: $(tmpDir) }
           - task: BatchScript@1
-            displayName: 'Validate Files'
-            timeoutInMinutes: 2
-            condition: succeededOrFailed()
             inputs:
               filename: 'azure-devops/validate-files.cmd'
               failOnStandardError: true
-              arguments: '$(buildOutputLocation)/validate/validate.exe'
+              arguments: '$(Build.SourcesDirectory)'
+            displayName: 'Validate Files'
+            timeoutInMinutes: 5
+            condition: succeededOrFailed()
             env: { TMP: $(tmpDir), TEMP: $(tmpDir) }
+          - task: Powershell@2
+            displayName: 'Create Diff'
+            inputs:
+              filePath: azure-devops/create-prdiff.ps1
+              arguments: '-DiffFile $(DiffFile)'
+              pwsh: false
+          - task: PublishBuildArtifacts@1
+            condition: failed()
+            displayName: 'Publish Format and Messages File Diff'
+            inputs:
+              PathToPublish: '$(DiffFile)'
+              ArtifactName: 'format.diff'

```

Comment:
When `validate` fails, its messages are still displayed in the Checks tab (which is good), but additionally this appears (if *only* `validate` failed), see #2682:

```
Publishing build artifacts failed with an error: Not found PathtoPublish: C:\a\1\a\format.diff
```

Can the publish task be configured to run only when clang-format has failed, causing `create-prdiff.ps1` to create a file? I believe that we can use [variables in conditions](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml) to achieve this.